src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+
+            internal readonly HashedWheelTimer timer;
+            internal readonly TimeSpan Deadline;
+
+            volatile int state = StInit;
+
+            // remainingRounds will be calculated and set by Worker.transferTimeoutsToBuckets() before the
+            // HashedWheelTimeout will be added to the correct HashedWheelBucket.
+            internal long RemainingRounds;
+
+            // This will be used to chain timeouts in HashedWheelTimerBucket via a double-linked-list.
+            // As only the workerThread will act on it there is no need for synchronization / volatile.
+            internal HashedWheelTimeout Next;
+
+            internal HashedWheelTimeout Prev;
+
+            // The bucket to which the timeout was added
+            internal HashedWheelBucket Bucket;
+
+            internal HashedWheelTimeout(HashedWheelTimer timer, ITimerTask task, TimeSpan deadline)
+            {
+                this.timer = timer;
+                this.Task = task;
+                this.Deadline = deadline;
+            }
+
+            public ITimer Timer => this.timer;
+
+            public ITimerTask Task { get; }
+
+            public bool Cancel()
+            {
+                // only update the state it will be removed from HashedWheelBucket on next tick.
+                if (!this.CompareAndSetState(StInit, StCancelled))
+                {
+                    return false;
+                }
+                // If a task should be canceled we put this to another queue which will be processed on each tick.
+                // So this means that we will have a GC latency of max. 1 tick duration which is good enough. This way
+                // we can make again use of our MpscLinkedQueue and so minimize the locking / overhead as much as possible.
+                this.timer.cancelledTimeouts.TryEnqueue(this);
+                return true;
+            }
+
+            internal void Remove()
+            {
+                HashedWheelBucket bucket = this.Bucket;
+                if (bucket != null)
+                {
+                    bucket.Remove(this);
+                }
+                else if (this.timer.ShouldLimitTimeouts())
+                {
+                    Interlocked.Decrement(ref this.timer.PendingTimeouts);
+                }
+            }
+
+            bool CompareAndSetState(int expected, int state)
+            {
+                return Interlocked.CompareExchange(ref this.state, state, expected) == expected;
+            }
+
+            internal int State => this.state;
+
+            public bool Cancelled => this.State == StCancelled;
+
+            public bool Expired => this.State == StExpired;
+
+            internal void Expire()
+            {
+                if (!this.CompareAndSetState(StInit, StExpired))
+                {
+                    return;
+                }
+
+                try
+                {
+                    this.Task.Run(this);
```

Comment:
correct; you can choose in your ITimeout implementation to schedule callback on a threadpool to avoid such issues. No limitations from HWT itself though.

---
In reply to: [114477953](https://github.com/Azure/DotNetty/pull/242#discussion_r114477953) [](ancestors = 114477953)